Fixes #14257- Enables provisioning of different trees with same relea…

parthaa opened the pull request:
…se/arch/version

Problem
--------
If you synced these 2 kickstart repos, which have the same
version/release/architecture we get into a mess with new hosts page.
RHEL Server 7.1 x86_64
RHEL Workstation 7.1 x86_64

Context:
--------
When we sync a kickstart repo from CDN we create an installation media ,
something that will look like ->
name :
Default_Organization/Library/Red_Hat_Server/Red_Hat_Enterprise_Linux_7_Server_Kickstart_x86_64_7_1
path: http://<sat
fqdn>/pulp/repos/Default_Organization/Library/content/dist/rhel/server/7/7.1/x86_64/kickstart

name:
Default_Organization/Library/Red_Hat_Workstation/Red_Hat_Enterprise_Linux_7_Workstation_Kickstart_x86_64_7_1
path: http://<sat
fqdn>/pulp/repos/Default_Organization/Library/content/dist/rhel/workstation/7/7.2/x86_64/kickstart/

To help further lets assume
Case 1
We just synced RHEL Server 7.1 x86_64  and not RHEL Workstation 7.1
x86_64

You go to the New Hosts page -> You select the following values
1) Lifecycle Environment -> Library
2) Content View -> Default Organization View
3) Content Source -> Default Capsule (actually any for that matter since
this is not properly filtered right now based of Env)
4) Under OS -> Arch -> x86_64
5) Operating System -> Red Hat 7.1

What happens now is this api call to
operatingsystems/<os_id>/available_kickstart_repo passing it the env,
cv, content source, arch , os
The response resolves it to 1 kickstart repo thats our synced Rhel 7.1
Kickstart. Something like

{
    "name": "Red Hat Enterprise Linux 7 Server Kickstart x86_64 7.1",
    "path": "http://<content source
url>/pulp/repos/Default_Organization/Library/content/dist/rhel/server/7/7.1/x86_64/kickstart"
}

The javascript on this page takes that one url and marks that as the
install media for foreman when provisioning the host.
The unattended kickstart template render also calls the same method
again to fetch this url

Now comes the actual bug
Case 2
We have both 7.1 and 7.1 workstation

In this case there are 2 available repos for the same values of env, cv,
content source, arch , os and here is where it becomes an issue. The
code right now picks the first repo matching this equation.

Fix:

This commit fixes this mess in the following way

1) When querying the server for the available KS repos we find more than
1 kickstart repo, we treat in the same way as if we found NO kickstart
repos. So this would mean that in the OS Tab => after the above steps
1-5 are completed you 'll see a Media box with all the Install Media
available in the org for that operating system as setup by foreman.

2) The user then selects the desired install media and submits
3) Then when generating the media URI for vm linuz the code replaces the
fqdn in the Install media path with the content source path if a content
source was selected (similar to the way it would if only one KS repo was
available).

parthaa commented:
To Test this

- Enable And Sync the right repos
![Enabled And Synced Repos](http://i.imgur.com/Kbvewq0.png)
- Assuming you have everything ready to provision the host, Goto New Hosts
- Setup the Lifecycle Env, Content View, Content Source
![New Host](http://i.imgur.com/QPWyRKQ.png)
- Fill up the Arch, OS and notice the new Media Options 
![Media Options](http://i.imgur.com/2BuSUDe.png)
- Select the media you want and Submit
- Make sure the URL used to pull contents is actually the content source and not the sat.
![Running VM](http://i.imgur.com/AzfvxYu.png)

- Also test the Hostgroups similarly
![New HostGroup](http://i.imgur.com/4kY8j8n.png)
- OS/Arch media options
![OS HostGroup](http://i.imgur.com/SkDqf40.png)

parthaa commented:
- Kickstarting on repos with a Unique arch/os/release works as normal -> http://i.imgur.com/hT9nY2v.png
![New Host](http://i.imgur.com/hT9nY2v.png )
- Note the way the media auto renders
![OS Host](http://i.imgur.com/OirIlEz.png)

mccun934 commented:
tested this and it worked well, chose the Workstation Media and provisioned properly.
![rwdtxgb](https://cloud.githubusercontent.com/assets/206457/14302407/3f638964-fb55-11e5-912f-dbfc2ffc0ce0.png)

ehelms commented on app/lib/katello/concerns/renderer_extensions.rb:
```diff
@@ -11,7 +11,8 @@ def kickstart_attributes_with_katello
         if @host.content_facet.try(:content_view) && @host.operatingsystem.is_a?(Redhat) &&
-          !@host.operatingsystem.kickstart_repo(@host).blank?
+           @host.operatingsystem.kickstart_repos(@host).size == 1 &&
+           @host.operatingsystem.kickstart_repos(@host).first.present?
```
Indent here since its a continuation

ehelms commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,31 @@ def construct_name(family)
+              path = repo.full_path(host.content_source)
+              path += "/" unless path[-1] == '/'
+              path.include?(uri_path)
+            end
```
It's not clear to me what this block is doing.

parthaa replied:
Not sure how to simplify this. Will add a comment

ehelms commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,31 @@ def construct_name(family)
       def medium_uri_with_content_uri(host, url = nil)
-        if host.try(:content_source) && (repo_details = kickstart_repo(host))
-          URI.parse(repo_details[:path])
+        if host.try(:content_source) && (repos = distribution_repositories(host)).present?
```
What about unit tests for this new situation?

parthaa replied:
@ehelms added

ehelms replied:
Are the repos returned by `distribution_repositories` not already guaranteed to be valid distribution repos for the host and content view? Before, there didn't appear to be any validation checking.

ehelms commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,31 @@ def construct_name(family)
+          distros.map do |distro|
+            {:name => distro.name, :path => distro.full_path(host.content_source)}
+          end
+        end
```
Should this warrant a unit test?

parthaa replied:
@ehelms added

ehelms replied:
Right now this would return nil if no distros or the host does not have a content source, should it return an empty array for consistency?

ehelms commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,63 @@ def construct_name(family)
+            #
+            # So we first get the Install Media path
+            # i.e http://<sat-fqdn>/pulp/repos/Default_Organization/Library/content/dist/rhel/workstation/7/7.2/x86_64/kickstart/
+            media_uri = medium_uri_without_content_uri(host, url)
```
If at this point we already have the media_uri, why do we need to do anything other than sub out the FQDN for the content source's FQDN?

parthaa replied:
We need to be sure that the selected media path is actually part of one of the kickstart repos. 

So for example if the user selects content source and the os has media associated to it that is not part of a ks repo we need to handle that. We just need to be sure.

ehelms commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,63 @@ def construct_name(family)
+              # no repo path was found for this matched, its quite possible the user chose a media that is not associated to
+              # any of our kickstart repos, even though its associated to the same os and accidentally chose a content source
+              # so just return the install media uri
+              media_uri
```
Does this need to call it the same as https://github.com/Katello/katello/pull/5940/files#diff-432a1f989a6588228bf10a570fbd46e2R93 ?

parthaa replied:
Yes that is the point..  Since we found no kickstart repo matching the media path, just assume that the user knows what they are doing and return  medium_uri_without_content_uri.  We do this few lines above this code.
```
 media_uri = medium_uri_without_content_uri(host, url)
```
So I thought of reusing that var instead of calling it again

parthaa commented:
@ehelms  I also updated the behavior of the available_kickstart_repo . I renamed it available_kickstart_repositories. I made it return [] instead of  nil and made the javascript handle it better. Look at https://github.com/Katello/katello/pull/5940/files#diff-dec6c572a895505beb991cdcaf69df36R165

ehelms commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,63 @@ def construct_name(family)
+          if repos.size == 1
+            # There is exactly 1 ks repo distribution matching the os, arch, release
+            # cv, env for this host. Just get the media path from available repo.
+            URI.parse(repos.first.full_path(host.content_source))
```
Why don't we also need to validate here that the distribution repository found matches the medium URI attached to the host? Seems like before we resting on an assumption that the UI or API would handle this properly. As I try to understand this better, I lean towards having some function that captures the "validation" logic so that both locations can make use of it.

parthaa replied:
Thats an interesting idea. But lemme explain something. Look at the image below
![OS Host](http://i.imgur.com/OirIlEz.png)
See how the media got auto populated.
So the UI interaction does not give the user a choice in selecting the media. If the user chose RedHat 7.2 as os, x86_64, content source, env, cv and there is only one matching media. So in theory we donot need to validate here.

parthaa commented:
Update Interactions To Test this

- Enable And Sync the right repos
![Enabled And Synced Repos](http://i.imgur.com/Kbvewq0.png)
- Assuming you have everything ready to provision the host, Goto New Hosts
- Setup the Lifecycle Env, Content View, Content Source
![New Host](http://i.imgur.com/QPWyRKQ.png)
- Fill up the Arch, OS and notice the new Media Options. You should a radio that will let you toggle between synced content and install media. The synced content must be a drop down that shows the Available Kickstart Distributions and the first one auto selected.
![Media Options](http://i.imgur.com/2Gc8UQh.png)
- Select the KS Repo you want and Submit
- Make sure the URL used to pull contents is actually the content source and not the sat.
![Running VM](http://i.imgur.com/AzfvxYu.png)

- Also test the Hostgroups similarly
![New HostGroup](http://i.imgur.com/4kY8j8n.png)
- OS/Arch media options
![OS HostGroup](http://i.imgur.com/SkDqf40.png)

Interactions to test 
- If you click on install media, does it correctly hide the synced content
- If you selected an install media and then click back on Synced Content does it correctly clear the selection
- If you then selected a KS repo and  Clicked "Install Media", is the Install media selection properly cleared
- If you unselected a content source or changed arch, os, cv, env is the behavior appropriate. If there is no available KS repos for your selections then Install Media should automatically get selected and Synced Content should be empty.
- If you had a host group with similar setup and you went to Hosts and asked it to use your selected host group does it behave in an expected manner.
- Can you assign multiple Ks distributions with the same OS, release, version to a content view (It should not raise any error)
- If your lifecycly/content view  had only one KS Distribution, does the drop down still show up properly with that one entry selected under synced content (It Should)

stbenjam commented on app/views/overrides/activation_keys/_host_media_type_select.html.erb:
```diff
@@ -0,0 +1,22 @@
+     <label class="col-md-2 control-label" for="use_install_media">Media Selection</label>
+     <div class="col-md-4">
+        <label class="radio-inline">
+          <input data-media-selector="true" id="host_use_install_media_true" name="media_selector" value="install_media" type="radio" <%= install_media_radio  %> /> <%= _('Install Media') %>
```
Hm, could we all this "All Media" or something, and put it after Synced Content? It's nitpicky but seems clearer to me.

```
[X] Synced Content [ ] All Media
```

Synced content is also technically install media.

parthaa replied:
updated 

![Media Options](http://i.imgur.com/vrFIqFU.png)

stbenjam replied:
Thanks, I like that a lot better :+1:

mccun934 commented:
some notes:

**1)** seeing duplicates on the Synced Content drop list

![ks-dupes](https://cloud.githubusercontent.com/assets/206457/14590089/408e9f26-04a7-11e6-9ae2-ae6b2474330f.png)

**2)** As a new user I'm not entire sure what the difference is from "Synced Content" and "All Media". Could we add a help tip or something next to the selector like we see under Partition Table , that explains the difference?

mccun934 commented:
if I clear my "Content Source" and have nothing selected shouldn't we grey out the "synced content" option?

Right now if I clear the Content Source, I'm still presented with both radio options but it doesn't re-render the drop list and only shows "All Media" options.

bbuckingham commented:
When selecting 'All Media', I am unable to see the entire media string.  It is not line wrapped, like it is for 'Synced Content'.   E.g.
    "Default_Organization/Library/Red_Hat_Server/Red_Hat_E"
    "Default_Organization/Library/Red_Hat_Workstation/Red_H"

bbuckingham commented:
After selecting Arch, OS, Synced Content, Media, if I change the OS, it correctly resets Media; however, I see it toggle the media selection as All Media and then Synced Content.  We may want to ensure that it just resets to All Media.  In one case (which I have not been able to repro), it set it to All Media and left it there.

bbuckingham commented:
Scenario: Create a hostgroup
- lifecycle env: dev
- content view: rhel7any (this view has 2 KS repos - Server & Workstation)
- OS: x86_64, RedHat 7.2
- media selection: Synced Content, selected the Server repo
- submit

Edit the hostgroup and observed that the selected media was the Workstation repo (which was the first in the list vs the expected Server repo)

bbuckingham commented:
Scenario: Create a hostgroup
- lifecycle env: dev
- content view: rhel7workstation
- OS: x86_64, RedHat 7.2
- media selection: All Media, selected Server repo
- submit

- create new host, select ^ hostgroup
- observe:  Media Selection, 'Synced Content' with Workstation repo selected

parthaa commented:
> if I clear my "Content Source" and have nothing selected shouldn't we grey out the "synced content" option?

This one is not a regression. I 'll try to see if I can fix it once I address other issues.

mccun934 commented:
tested a build of this one and it worked almost perfectly. The only issue I hit were scenarios where I selected an OS that had 0 matches , say:

Select RHEL 6.5 when I had no synced RHEL 6.5.

In this case if you switch from that to say, RHEL 6.7 which I did have synced, you end up with the double select lists:

![ss-1123](https://cloud.githubusercontent.com/assets/206457/14627572/8f5443d2-05aa-11e6-94f7-1081c6744e0c.png)

other than that, worked fine. So, go from a blank -> populated and you get the double.

mccun934 commented:
retested with the latest update, my problem with the double select lists is gone.

provisioned EL 7.2 Server and Atomic Host without error.

ACK for me from a user perspective

bbuckingham commented:
Scenario: Create a 2 Host Groups and then a New Host

- create 2 Host Groups (e.g. groupA with 'Synced Content' Workstation and groupB with 'Synced Content' Server)
- create a new host
- select Host Group: groupA
- observe that OS/Arch/Synced Content are correct (e.g. pointing to Workstation)
- change the Host Group to groupB
- observe that Synced Content is blank.   I assume at this point, it should have been updated to Server

parthaa commented:
> Scenario: Create a 2 Host Groups and then a New Host

This scenario should now be fixed with my last commit.

bbuckingham commented:
@parthaa, Thanks for the update.  Got another strange one related to:

Scenario: Create 2 Host Groups and then a New Host

Now, when changing from groupA to groupB, the 'Synced Content' is getting updated properly, but the Content View is not.

parthaa commented:
> Now, when changing from groupA to groupB, the 'Synced Content' is getting updated properly, but the Content View is not.

Not a regression. Definitely a bug .

parthaa commented:
> Not a regression. Definitely a bug .

That being said.. Also fixed ...

parthaa commented:
@bbuckingham applied the changes to the test sat.

jlsherrill commented on app/helpers/katello/hosts_and_hostgroups_helper.rb:
```diff
@@ -98,5 +115,37 @@ def content_views_for_host(host, options)
+        view_options = new_host.operatingsystem.kickstart_repos(new_host).map { |repo| OpenStruct.new(repo) }
+      elsif ["operatingsystem_id", 'content_view_id',
+             'lifecycle_environment_id', 'content_source_id',
+             'architecture_id'].all? { |key| !params[key].blank? }
```
This is a tad ugly, mind moving this up above the if statement to a variable?

parthaa replied:
So this method gets called in 2 places

1. On initial page load or a host group selection. At that point the host object is already populated and we should just use that.
1. Once you chose a diff os/content source/arch/lifecycle env/cv

In case  2 we want it to play by the rules of "one of these params" and in case 1 we want it to behave as if everything is already set right and we just need the KS repos ..

jlsherrill replied:
thats fine, i'm just saying the 3 line elsif with a block is a bit ugly, can you move that to the start of the method?  Sure it'll be executed needlessly in some cases but this not exactly a compute intensive check

jlsherrill replied:
or maybe as a separate smaller method that is called from the elsif

jlsherrill commented on app/models/katello/concerns/host_managed_extensions.rb:
```diff
@@ -10,7 +10,7 @@ module HostManagedExtensions
         alias_method_chain :info, :katello
         alias_method_chain :smart_proxy_ids, :katello
-
+        belongs_to :kickstart_repository, :class_name => "::Katello::Repository", :foreign_key => :kickstart_repository_id
```
should this be in the content_facet?

jlsherrill replied:
note if its too complicated due to differences with hostgroup, thats okay, just wanted to pose the question

parthaa replied:
I was about to say yes on Content Facet, except that we'd have to deal specially for host vs hostgroup..

parthaa replied:
@jlsherrill I did work on moving to content facet, but pushed it to a diff branch because I was not comfortable testing everything this late in the game.. Look at the diff here https://github.com/parthaa/katello/commit/35b6ead6bf30b02388e9f56eb844b41371d6d57e and tell me what you think..

jlsherrill replied:
I'm fine with those changes, if brad doesn't mind re-testing with those changes i'd be okay with them.

jlsherrill commented on app/models/katello/concerns/redhat_extensions.rb:
```diff
@@ -44,16 +44,22 @@ def construct_name(family)
       def medium_uri_with_content_uri(host, url = nil)
-        if host.try(:content_source) && (repo_details = kickstart_repo(host))
-          URI.parse(repo_details[:path])
+        if  host.try(:content_source) && host.kickstart_repository.present?
```
Extra whitespace between if and host?

jlsherrill commented on app/helpers/katello/hosts_and_hostgroups_helper.rb:
```diff
@@ -98,5 +115,37 @@ def content_views_for_host(host, options)
+        end
+      end
+      view_options
+    end
```
Can you write a few tests for this helper?

jlsherrill replied:
This is partially replacing the UI action which was well tested :)

parthaa replied:
Any suggestions where to look for these tests?

parthaa replied:
@jlsherrill added new helper tests. Thank you for the suggestion

bbuckingham commented:
@parthaa, Thanks for the update. Another variation...

Scenario: Create 2 Host Groups and then a New Host - this time using 'All Media' (instead of 'Synced Content')

- groupA - All Media, Media=Workstation
- groupB - All Media, Media=Server
- create a new host
- select Host Group: groupA
- observe that OS/Arch/Media are correct (e.g. pointing to Workstation)
- change the Host Group to groupB
- observe that Media is unchanged.  It is set to Workstation, but should now be Server.

parthaa commented:
> Scenario: Create 2 Host Groups and then a New Host - this time using 'All Media' (instead of 'Synced Content')
Good catch 

Just verified that this is not a regression. It needs to be fixed in foreman if at all. I believe it applies the host group completely only the first time for the medium and once host is set to it (even if not saved) it does not let you change. I saw the same behaviour with OS. I had 2 host groups with different OSes and once i set the first one it wont let me change it.

bbuckingham commented:
@parthaa, If that last scenario is in foreman core, we can definitely track that last item as a separate redmine.  Based on that, I didn't see anything else from a user point of view.  Nice work!

Referring to:

    Scenario: Create 2 Host Groups and then a New Host - this time using 'All Media' (instead of 'Synced   Content')

bbuckingham commented:
Tested with the above updates on a patched lab environment.  The behavior is what is expected per our conversations; therefore, I am ok with moving forward with this PR and having issues created to address our known gaps, which @parthaa can best summarize.  (e.g. On a 'New Host', if the user changes the selected host group (e.g. from group A to B) it will not propagate the Operating System tab details from group B to the 'new host'.)

parthaa commented:
@bbuckingham also created http://projects.theforeman.org/issues/14737 and  http://projects.theforeman.org/issues/14699 to be fixed in foreman...

parthaa commented:
[test]

bbuckingham commented:
Scenario 1:  Creating a hostgroup where CV has multiple distributions

- LIfecycle Environment : select dev
- Content View: select rhel7any (which has the Server and Workstation KS repos)
- Content Source: select default (i.e the katello server)
- Operating system: Arch: x86_64, OS: RedHat 7.2
- Media Selection: Synced Content
- Synced Content: observe 2 media exist, selected the second one (Server in my case)
- Submit

- Go back in to view the HG created and see that the Synced Content is set to 'Workstation' (and not Server)

bbuckingham commented:
Scenario 2: Creating a hostgroup where select a Lifecycle Env & View, but want to use 'Install Media'
- LIfecycle Environment : select dev
- Content View: select rhel7workstation (which has only the Workstation KS repo)
- Content Source: select default (i.e the katello server)
- Operating system: Arch: x86_64, OS: RedHat 7.2
- Media Selection: Install Media
- Synced Content: observe 2 media exist, selected the first one (Server in my case)
- Submit
- Go back in to view the HG created and Operating System tab indicates that 'Synced Content' is selected and the 'Workstation'  media.  (Note: was expecting 'Install Media' with Server selected)

bbuckingham commented:
Scenario 3:  Hosts -> New Host

```
undefined method `kickstart_repository_id' for #<Host::Managed:0x007f54cc099d90>

app/helpers/form_helper.rb:136:in `block in select_f'
app/helpers/form_helper.rb:344:in `block (3 levels) in field'
app/helpers/form_helper.rb:340:in `block (2 levels) in field'
app/helpers/form_helper.rb:338:in `block in field'
app/helpers/form_helper.rb:337:in `field'
app/helpers/form_helper.rb:133:in `select_f'
app/views/common/os_selection/_operatingsystem.html.erb:28:in `block in _a555cbceb377d0095c5452a55e2138df'
app/views/common/os_selection/_operatingsystem.html.erb:1:in `_a555cbceb377d0095c5452a55e2138df'
app/views/hosts/provision_method/build/_form.html.erb:6:in `_16a0369183d8fee27f0b892eb880cde0'
app/views/hosts/_operating_system.html.erb:20:in `block in _28610d74b0a7f9aa9c41569843bc352a'
app/views/hosts/_operating_system.html.erb:18:in `each'
app/views/hosts/_operating_system.html.erb:18:in `_28610d74b0a7f9aa9c41569843bc352a'
app/views/hosts/_unattended.html.erb:6:in `_80c76dc8fc8da3545c192f1ecd82c2e7'
app/views/hosts/_form.html.erb:142:in `block (2 levels) in _a511d3f2a7073203d0a897d58cea1105'
...
...
```

bbuckingham commented:
@parthaa, Let me reset my environment.  The code in my branch does not look correct.  I'll remove the above comments, if they do not occur again or let you know if they were recreated.   Stand-by...

bbuckingham commented:
@parthaa, I mentioned in IRC, but I just finished testing the PR.  I ran numerous usability tests on my dev environment which was cleanly patched with the PR changes.  From there, I tested several scenarios involving hostgroups and hosts.  I then moved to the lab machine that you patched and tested out provisioning of a couple of VMs.  Everything behaved as we expect it to.  From my point of view, ACK.

mccun934 commented:
:fireworks:
